presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDiskPageInput.java

```diff
@@ -0,0 +1,148 @@
+    private int currentIteratorIndex;
+
+    public PrestoSparkDiskPageInput(
+            PagesSerde pagesSerde,
+            List<List<PrestoSparkDiskPage>> broadcastTableFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext,
+            PrestoSparkBroadcastTableCacheManager prestoSparkBroadcastTableCacheManager,
+            StageId stageId)
+    {
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.serializedPageIterators = broadcastTableFilesInfo.stream()
+                .map(tableFiles -> loadBroadcastTable(tableFiles, tempStorage, tempDataOperationContext, prestoSparkBroadcastTableCacheManager, stageId))
+                .collect(Collectors.toList());
+    }
+
+    private synchronized Iterator<PrestoSparkSerializedPage> loadBroadcastTable(
+            List<PrestoSparkDiskPage> broadcastTaskFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext,
+            PrestoSparkBroadcastTableCacheManager prestoSparkBroadcastTableCacheManager,
+            StageId stageId)
+    {
+        // Empty broadcast table
+        if (broadcastTaskFilesInfo.size() == 0) {
+            return Collections.emptyIterator();
+        }
+
+        List<PrestoSparkSerializedPage> pages = new ArrayList<>();
+
+        // Try to get table from cache
+        if (prestoSparkBroadcastTableCacheManager.isBroadcastTableCachedForStage(stageId)) {
+            pages = prestoSparkBroadcastTableCacheManager.getBroadcastTableForStage(stageId);
+        }
+        else {
+            try {
+                shuffle(broadcastTaskFilesInfo);
+                for (PrestoSparkTaskOutput taskFileInfo : broadcastTaskFilesInfo) {
+                    PrestoSparkDiskPage prestoSparkDiskPage = (PrestoSparkDiskPage) taskFileInfo;
+                    TempStorageHandle tempStorageHandle = tempStorage.deserialize(prestoSparkDiskPage.getSerializedStorageHandle());
+                    log.info("Reading path: " + tempStorageHandle.toString());
+                    InputStream inputStream = tempStorage.open(tempDataOperationContext, tempStorageHandle);
+                    Iterator<SerializedPage> pagesIterator = readSerializedPages(new InputStreamSliceInput(inputStream));
+                    int pageIndex = 0;
+                    while (pagesIterator.hasNext()) {
+                        SerializedPage page = pagesIterator.next();
+                        long checksum = computeChecksum(page.getSlice().byteArray());
+                        if (prestoSparkDiskPage.getPageChecksum().get(pageIndex) != checksum) {
+                            throw new RuntimeException("Disk page checksum does not match. " +
+                                    "Data seems to be corrupted on disk for file " + tempStorageHandle.toString());
+                        }
+                        pages.add(toPrestoSparkSerializedPage(page));
```

Comment:
Ideally it would be best to cache already deserialized pages. Though deserialization should be done outside the lock. Only reading pages from the input stream must be done under the lock.